File: C++/circularQueue.cpp
#include<iostream>
#include<vector>
using namespace std;

class circularQueue {
private:
    int front;
    int rear;
    int capacity;
    int size;
    int *queue_array;

public:
    // constructor
    circularQueue(int capacity): front(-1), rear(-1), capacity(capacity), size(0) {
        if (capacity <= 0) {
            throw invalid_argument("Capacity must be positive");
        }
        queue_array = new int[capacity];
    }

    // method declarations
    void enqueue(int data);
    void dequeue();
    int peek();
    bool empty();
    bool full();
    void display();
    void reset();

    // destructor
    ~circularQueue() {
        cout<<"Destructor called | ";
        reset();
        delete queue_array;
        cout<<"Resources cleared"<<endl;
    }
};

void circularQueue::enqueue(int data) {
    if(full()) {
        cout<<"Queue overflow, cannot enqueue!"<<endl;
        return;
    }

    rear = (rear + 1) % capacity;
    if(empty()) front = rear;

    queue_array[rear] = data;
    size += 1;
    cout<<data<<" enqueued successfully to the rear of the queue"<<endl;
    return;
}

void circularQueue::dequeue() {
    if(empty()) {
        cout<<"Queue underflow, cannot dequeue"<<endl;
        return;
    }

    int x = queue_array[front];
    if(front == rear) front = -1;
    front = (front + 1) % capacity;
    size -= 1;
    cout<<x<<" dequeued from the front of the queue"<<endl;
    return;
}

int circularQueue::peek() {
    if(empty()) {
        cout<<"Queue underflow, cannot peek"<<endl;
        return -1;
    }
    
    int x = queue_array[front];
    cout<<"Element peeked at the front: "<<x<<endl;
    return x;
}

bool circularQueue::empty() {
    return size == 0;
}

bool circularQueue::full() {
    return size == capacity;
}

void circularQueue::display() {
    cout<<"Queue elements: ";
    int r = front;
    int i = 0;
    while(i < size){
        cout<<queue_array[r]<<" ";
        r = (r+1) % capacity;
        i += 1;
    }
    cout<<endl;
}

void circularQueue::reset() {
    front = -1;
    rear = -1;
    size = 0;
    cout<<"Queue reset"<<endl;
}

int main() {

    circularQueue queue = circularQueue(5);

    if(queue.empty()) cout<<"Queue is empty"<<endl;
    if(queue.full()) cout<<"Queue is full"<<endl;

    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);
    queue.enqueue(4);
    queue.enqueue(5);
    queue.enqueue(6);

    queue.display();

    queue.dequeue();
    queue.dequeue();
    queue.dequeue();
    queue.display();


    queue.peek();

    queue.enqueue(10);
    queue.enqueue(20);
    queue.enqueue(30);

    queue.display();

    queue.dequeue();
    queue.dequeue();

    queue.enqueue(100);
    queue.enqueue(200);
    queue.enqueue(300);

    queue.display();

    queue.dequeue();
    queue.dequeue();
    queue.dequeue();
    queue.dequeue();
    queue.dequeue();

    queue.peek();

    queue.enqueue(1000);
    queue.enqueue(2000);
    queue.enqueue(3000);

    queue.display();
    queue.reset();

    return 0;
}

File: Go/CircularQueue/circularQueue.go
package main

import (
	"fmt"
	"errors"
)

type circularQueue struct {
	front int
	rear int
	capacity int
	size int
	queue_array []int
}


func InitCircularQueue(capacity int) (*circularQueue, error) {
	if (capacity <= 0) {
		return nil, errors.New("Capacity must be positive, cannot create circular queue")
	}
	var queue *circularQueue = &circularQueue{}
	queue.front = -1
	queue.rear = -1
	queue.capacity = capacity
	queue.size = 0
	queue.queue_array = make([]int, capacity)
	return queue, nil
}


func (queue *circularQueue) Enqueue(data int) {
	if (queue.IsFull()) {
		fmt.Println("Queue overflow, cannot enqueue!")
		return
	}

	queue.rear = (1 + queue.rear) % queue.capacity
	queue.queue_array[queue.rear] = data

	if (queue.IsEmpty()) {
		queue.front = queue.rear
	}
	queue.size += 1
	fmt.Print(data, " enqueued to the rear of the queue\n")
	return
}

func (queue *circularQueue) Dequeue() {
	if (queue.IsEmpty()) {
		fmt.Println("Queue underflow, cannot dequeue")
		return
	}

	var x int = queue.queue_array[queue.front]
	if (queue.front == queue.rear) {
		queue.front = -1
	}
	queue.front = (queue.front + 1) % queue.capacity
	queue.size -= 1
	fmt.Print(x, " dequeued from the front of the queue\n")
	return
}

func (queue *circularQueue) Peek() (int, error) {
	if(queue.IsEmpty()) {
		return -1, errors.New("Queue underflow, cannot peek")
	}
	var x int = queue.queue_array[queue.front]
	return x, nil
}

func (queue *circularQueue) IsEmpty() bool {
	return queue.size == 0
}

func (queue *circularQueue) IsFull() bool {
	return queue.size == queue.capacity
}

func (queue *circularQueue) Display() {
	if(queue.IsEmpty()) {
		fmt.Println("Queue underflow, nothing to display")
		return
	}

	fmt.Print("Queue elements (front -> rear): ")
	for x,i := 0, queue.front; x < queue.size; i, x = (i+1)%queue.capacity, x + 1 {
		fmt.Print(queue.queue_array[i], " ")
	}
	fmt.Println()
}

func (queue *circularQueue) Clear() {
	
	queue.queue_array = nil
	queue.front = -1
	queue.rear = -1
	queue.size = 0
	fmt.Println("Queue cleared")
	return 
}

func main() {
	queue, err := InitCircularQueue(5)
	if (err != nil) {
		fmt.Println("Error: ", err)
		return
	}
	fmt.Println("Initial size: ", queue.size, " | Initial capacity: ", queue.capacity)

	queue.Enqueue(1)	
	queue.Enqueue(2)	
	queue.Enqueue(3)	
	queue.Enqueue(4)
	queue.Enqueue(5)
	queue.Enqueue(6)

	queue.Display()

	queue.Dequeue()
	queue.Dequeue()
	queue.Dequeue()

	queue.Display()
	x, err := queue.Peek()
	if(err != nil) {
		fmt.Println("Error: ", err)
	} else {
		fmt.Println("Peek element: ", x)
	}

	queue.Dequeue()
	queue.Dequeue()
	queue.Dequeue()

	queue.Display()
	x, err = queue.Peek()
	if(err != nil) {
		fmt.Println("Error: ", err)
	} else {
		fmt.Println("Peek element: ", x)
	}

	queue.Enqueue(10)	
	queue.Enqueue(20)	

	queue.Dequeue()

	queue.Enqueue(30)	
	queue.Enqueue(40)	

	queue.Dequeue()
	queue.Dequeue()

	queue.Display()
	x, err = queue.Peek()
	if(err != nil) {
		fmt.Println("Error: ", err)
	} else {
		fmt.Println("Peek element: ", x)
	}

	queue.Clear()
}